core/src/Streamly/Internal/FileSystem/Path.hs

```diff
@@ -1,515 +1,194 @@
+-- could refer to a segment or segments.
 --
--- /Unimplemented/
-file :: QuasiQuoter
-file = undefined
-
--- | Generates an @Abs (Dir Path)@ type from an interpolated string literal.
+-- A path that refers to a particular object in the file system  is called a
+-- location e.g. /usr is a location, . is a location, ./bin is a location. A
+-- location could be absolute e.g. /usr or it could be relative e.g. ./bin . A
+-- location always has two components, a specific "root" which could be
+-- explicit or implicit, and a path segment relative to the root. A location
+-- with a fixed root is known as an absolute location whereas a location with
+-- an implicit root e.g. "./bin" is known as a relative location.
 --
--- /Unimplemented/
-absdir :: QuasiQuoter
-absdir = undefined
-
--- | Generates an @Rel (Dir Path)@ type from an interpolated string literal.
+-- A path that does not refer to a particular location but defines steps to go
+-- from some place to another is a path segment. For example, "local/bin" is a
+-- path segment whereas "./local/bin" is a location.
 --
--- /Unimplemented/
-reldir :: QuasiQuoter
-reldir = undefined
-
--- | Generates an @Abs (File Path)@ type from an interpolated string literal.
+-- Locations can never be appended to another location or to a path segment
+-- whereas a segment can be appended.
 --
--- /Unimplemented/
-absfile :: QuasiQuoter
-absfile = undefined
-
--- | Generates an @Rel (File Path)@ type from an interpolated string literal.
+-- == Comparing Paths
 --
--- /Unimplemented/
-relfile :: QuasiQuoter
-relfile = undefined
-
-------------------------------------------------------------------------------
--- Statically Verified Strings
-------------------------------------------------------------------------------
-
--- | Generates a 'Path' type.
+-- We can compare two absolute locations or path segments but we cannot compare
+-- two relative locations. If each component of the path is the same then the
+-- paths are considered to be equal.
 --
--- /Unimplemented/
-mkPath :: String -> Q Exp
-mkPath = undefined
-
--- | Generates an @Abs Path@ type.
+-- == Implicit Roots (.)
 --
--- /Unimplemented/
-mkAbs :: String -> Q Exp
-mkAbs = undefined
-
--- | Generates an @Rel Path@ type.
+-- On Posix and Windows "." implicitly refers to the current directory. On
+-- Windows a path like @/Users/@ has the drive reference implicit. Such
+-- references are contextual and may have different meanings at different
+-- times.
 --
--- /Unimplemented/
-mkRel :: String -> Q Exp
-mkRel = undefined
-
--- | Generates an @Dir Path@ type.
+-- @./bin@ may refer to a different location depending on what "." is
+-- referring to. Thus we should not allow @./bin@ to be appended to another
+-- path, @bin@ can be appended though. Similarly, we cannot compare @./bin@
+-- with @./bin@ and say that they are equal because they may be referring to
+-- different locations depending on in what context the paths were created.
```

Comment:
The same can be said for `..`?

Reply:
No. `..` means go one node up from wherever it is used, it is not necessarily from the current directory e.g. `a/../b`. it is always used as a way to specify a segment and never a location.